### PR TITLE
[Feature] air with additional fields

### PR DESCRIPTION
**Pull request by vg-svitla**

### CODE REVIEW

File: Office 365/o365/ingest/parser.yml [hunk 1]
```diff
@@ -321,6 +319,20 @@ stages:
...
+          rule.name: '{{parse_data.ParsedData.AlertDisplayName}}'
+
       - set:
           event.kind: '{% if parse_data.ParsedData.IsIncident %}{{"alert"}}{% else %}{{"event"}}{% endif %}'
+
+          ## Here is a problem with types, because all of this fields should be lists but in our definition they are strings
+          event.url: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Urls}} {% endfor %}'
+#          source.ip: '{% for Property in  parse_data.ParsedData.Entities %} {{Property.SenderIP}} {% endfor %}'
+          email.subject: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Subject}} {% endfor %}'
+          source.domain: '{% for Property in parse_data.ParsedData.Entities %} {{Property.P1SenderDomain}} {% endfor %}'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:335]: [`email.from.address`](https://www.elastic.co/guide/en/ecs/current/ecs-email.html#field-email-from-address) seems more suitable. Transform the result as an array.
```suggestion
          email.from.address: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.P1SenderDomain}}", {% endfor %}]'
```

File: Office 365/o365/ingest/parser.yml [hunk 2]
```diff
@@ -321,6 +319,20 @@ stages:
...
+          threat.enrichments: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Threats}} {% endfor %}'
+          email.sender.address: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Sender}} {% endfor %}'
+          destination.user.email: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Recipient}} {% endfor %}'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:338]: [`email.to.address`](https://www.elastic.co/guide/en/ecs/current/ecs-email.html#field-email-to-address) seems more suitable. Transform as an array as expected by the field
```suggestion
          email.to.address: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Recipient}}", {% endfor %}]'
```

File: Office 365/o365/ingest/parser.yml [hunk 3]
```diff
@@ -321,6 +319,20 @@ stages:
...
       - set:
           event.kind: '{% if parse_data.ParsedData.IsIncident %}{{"alert"}}{% else %}{{"event"}}{% endif %}'
+
+          ## Here is a problem with types, because all of this fields should be lists but in our definition they are strings
+          event.url: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Urls}} {% endfor %}'
+#          source.ip: '{% for Property in  parse_data.ParsedData.Entities %} {{Property.SenderIP}} {% endfor %}'
+          email.subject: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Subject}} {% endfor %}'
+          source.domain: '{% for Property in parse_data.ParsedData.Entities %} {{Property.P1SenderDomain}} {% endfor %}'
+          threat.enrichments: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Threats}} {% endfor %}'
+          email.sender.address: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Sender}} {% endfor %}'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:337]: According to the tests, these fields are not suitable with the generated values. I suggest creating custom fields that accept an array.

File: Office 365/o365/ingest/parser.yml [hunk 4]
```diff
@@ -321,6 +319,20 @@ stages:
...
+          log.level: '{{parse_data.ParsedData.Severity}}'
+          rule.name: '{{parse_data.ParsedData.AlertDisplayName}}'
+
       - set:
           event.kind: '{% if parse_data.ParsedData.IsIncident %}{{"alert"}}{% else %}{{"event"}}{% endif %}'
+
+          ## Here is a problem with types, because all of this fields should be lists but in our definition they are strings
+          event.url: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Urls}} {% endfor %}'
+#          source.ip: '{% for Property in  parse_data.ParsedData.Entities %} {{Property.SenderIP}} {% endfor %}'
+          email.subject: '{% for Property in parse_data.ParsedData.Entities %} {{Property.Subject}} {% endfor %}'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:334]: According to the tests, these fields are not suitable with the generated values. I suggest creating custom fields that accept an array.

**Review by squioc — CHANGES_REQUESTED** [review]

File: Office 365/o365/_meta/fields.yml [hunk 5]
```diff
@@ -343,3 +343,13 @@ email.attachments:
...
+
+email.subjects:
+  name: email.subjects
+  description: "A list of email subjects"
+  type: array
+
+email.from.domain:
+  name: email.from.domain
+  description: "The domain of the sender. Might be an array"
+  type: keyword
```
Comment (squioc) [inline Office 365/o365/_meta/fields.yml:355]: Custom fields must not look similar to ECS one. They must be prefixed with the name of the vendor and the name of the format (see [Taxonomy](https://github.com/SEKOIA-IO/intake-formats/blob/main/doc/taxonomy.md#custom-taxomony))
```suggestion
office365.investigation.email.subjects:
  name: office365.investigation.email.subjects
  description: "A list of email subjects"
  type: array

office365.investigation.email.sender.domains:
  name: office365.investigation.email.sender.domains
  description: "The domain of the sender. Might be an array"
  type: keyword
```

(Do not account of the `email.*` fields declared in this taxonomy; they date from a time the ECS taxonomy was not up-to-date).

File: Office 365/o365/ingest/parser.yml [hunk 6]
```diff
@@ -320,6 +318,21 @@ stages:
...
           event.kind: '{% if parse_data.ParsedData.IsIncident %}{{"alert"}}{% else %}{{"event"}}{% endif %}'
+
+      - set:
+          event.url: '[{% for Property in parse_data.ParsedData.Entities %}{% for Url in Property.Urls %} "{{Url}}", {% endfor %}{% endfor %}]'
+          source.ip: '[{% for Property in  parse_data.ParsedData.Entities %} "{{Property.SenderIP}}", {% endfor %}]'
+          email.subjects: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Subject}}", {% endfor %}]'
+          email.from.domain: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.P1SenderDomain}}", {% endfor %}]'
+          email.from.address: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Sender}}", {% endfor %}]'
+          threat.enrichments: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Threats}}", {% endfor %}]'
+          destination.user.email: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Recipient}}", {% endfor %}]'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:337]:
```suggestion
          email.to.address: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Recipient}}", {% endfor %}]'
```

File: Office 365/o365/ingest/parser.yml [hunk 7]
```diff
@@ -320,6 +318,21 @@ stages:
...
       - set:
           event.kind: '{% if parse_data.ParsedData.IsIncident %}{{"alert"}}{% else %}{{"event"}}{% endif %}'
+
+      - set:
+          event.url: '[{% for Property in parse_data.ParsedData.Entities %}{% for Url in Property.Urls %} "{{Url}}", {% endfor %}{% endfor %}]'
+          source.ip: '[{% for Property in  parse_data.ParsedData.Entities %} "{{Property.SenderIP}}", {% endfor %}]'
+          email.subjects: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Subject}}", {% endfor %}]'
+          email.from.domain: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.P1SenderDomain}}", {% endfor %}]'
+          email.from.address: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Sender}}", {% endfor %}]'
+          threat.enrichments: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Threats}}", {% endfor %}]'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:336]: [`threat.enrichments`](https://www.elastic.co/guide/en/ecs/8.6/ecs-threat.html#field-threat-enrichments) expects a list of [`indicators`](https://www.elastic.co/guide/en/ecs/8.6/ecs-threat.html#field-threat-enrichments-indicator).
I think the values are not suitable for this field. I suggest creating a custom field `office365.investigation.threats`.

**Review by squioc — CHANGES_REQUESTED** [review]

File: Office 365/o365/ingest/parser.yml [hunk 8]
```diff
@@ -320,6 +318,21 @@ stages:
...
+          event.end: "{{json_event.message.EndTimeUtc}}"
           host.name: "{{parse_data.ParsedData.MachineName}}"
+          log.level: "{{parse_data.ParsedData.Severity}}"
+          rule.name: "{{parse_data.ParsedData.AlertDisplayName}}"
+
       - set:
           event.kind: '{% if parse_data.ParsedData.IsIncident %}{{"alert"}}{% else %}{{"event"}}{% endif %}'
+
+      - set:
+          event.url: '[{% for Property in parse_data.ParsedData.Entities %}{% for Url in Property.Urls %} "{{Url}}", {% endfor %}{% endfor %}]'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:331]:
```suggestion
          office365.investigation.email.urls: '[{% for Property in parse_data.ParsedData.Entities %}{% for Url in Property.Urls %} "{{Url}}", {% endfor %}{% endfor %}]'
```

File: Office 365/o365/tests/automated_investigation_and_response.json [hunk 9]
```diff
@@ -69,16 +70,38 @@
           "correlation_key": "correlation_key_value",
           "category": "category_value",
           "source_type": "source_alert_type_value"
+        },
+        "threats": [],
+        "email": {
+          "subjects": [],
+          "sender": {
+            "domains": []
+          }
```
Comment (squioc) [inline Office 365/o365/tests/automated_investigation_and_response.json:79]: empty lists

File: Office 365/o365/ingest/parser.yml [hunk 10]
```diff
@@ -320,6 +318,21 @@ stages:
...
+          source.ip: '[{% for Property in  parse_data.ParsedData.Entities %} "{{Property.SenderIP}}", {% endfor %}]'
```
Comment (squioc) [inline Office 365/o365/ingest/parser.yml:332]: `source.ip` is not a list. I suggest creating a custom field.
```suggestion
          office365.investigation.email.sender.ip: '[{% for Property in  parse_data.ParsedData.Entities %} "{{Property.SenderIP}}", {% endfor %}]'
```

**Review by squioc — CHANGES_REQUESTED** [review]

File: Office 365/o365/ingest/parser.yml [hunk 11]
```diff
@@ -320,6 +318,21 @@ stages:
...
+          office365.investigation.email.subjects: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Subject}}", {% endfor %}]'
+          office365.investigation.email.sender.domains: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.P1SenderDomain}}", {% endfor %}]'
+          office365.investigation.threats: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Threats}}", {% endfor %}]'
+          email.from.address: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Sender}}", {% endfor %}]'
+          email.to.address: '[{% for Property in parse_data.ParsedData.Entities %} "{{Property.Recipient}}", {% endfor %}]'
+        filter: "{{parse_data.ParsedData.Entities != None and parse_data.ParsedData.Entities != []}}"
```
[the remark this answers is not in the document]
Comment (vg-svitla) [reply Office 365/o365/ingest/parser.yml]: Yes very good point thank you!

**Review by squioc — APPROVED** [review]

**Pull request merged**